### PR TITLE
mnemonic support added to desktop app

### PR DESCRIPTION
**Pull request by ozgunozerk**
This PR extracts the mnemonic from the farmer lib from the tauri backend, and uses it in the desktop app

### CODE REVIEW

File: src-tauri/src/main.rs [hunk 1]
```diff
@@ -20,22 +21,27 @@ use tauri::{
...
-    public_key.unwrap()
+async fn farming(path: String) -> FarmerIdentity {
+    let farmer_identity = farm(path.into(), "ws://127.0.0.1:9944").await;
+    farmer_identity.unwrap()
 }
 
 #[tauri::command]
 fn get_disk_stats(dir: String) -> DiskStats {
-    info!("{}", dir);
+    println!("{}", dir);
```
Comment (nazar-pc) [inline src-tauri/src/main.rs:44]: Did you actually mean to do this? Looks like a debug message to me at best.

File: src-tauri/src/main.rs [hunk 2]
```diff
@@ -195,15 +201,21 @@ pub(crate) async fn farm(
...
         Ok(())
     });
 
-    Ok(identity.public_key().to_bytes())
+    Ok(FarmerIdentity {
+        public_key: identity.public_key().to_bytes(),
+        mnemonic: Mnemonic::from_entropy(identity.entropy(), Language::English)
+            .unwrap()
+            .phrase()
+            .parse()?,
```
Comment (nazar-pc) [inline src-tauri/src/main.rs:219]: Parsing `&str` to get `String` is non-idiomatic and I think in this case not needed:
```suggestion
        mnemonic: Mnemonic::from_entropy(identity.entropy(), Language::English)
            .unwrap()
            .into_phrase(),
```

File: src-tauri/src/main.rs [hunk 3]
```diff
@@ -195,15 +201,21 @@ pub(crate) async fn farm(
...
-          },
-          res = farming_instance.wait() => if let Err(error) = res {
-            return Err(anyhow!(error)) // TODO: connect this error to frontend, or log it
-          },
+            res = plotting_instance.wait() => if let Err(error) = res {
+                return Err(anyhow!(error))
+            },
+            res = farming_instance.wait() => if let Err(error) = res {
+                return Err(anyhow!(error))
+            },
```
Comment (nazar-pc) [inline src-tauri/src/main.rs:209]: Why were TODOs removed?

File: src/lib/client.ts [hunk 4]
```diff
@@ -63,6 +63,11 @@ export interface ClientData {
   farming: ClientFarming
 }
 
+export interface ClientIdentity {
+  public_key: string,
```
Comment (nazar-pc) [inline src/lib/client.ts:67]: It is non-idiomatic to use `snake_case` in TypeScript, I would suggest to add `#[serde(rename_all = "snake_case")]` to the struct on backend.

File: src/lib/client.ts [hunk 5]
```diff
@@ -203,6 +208,8 @@ export class Client {
   }
 }
 
-export function startFarming(path: string): Promise<string> {
-  return (tauri.invoke('farming', { path })).then(value => value as string)
+export async function startFarming(path: string): Promise<ClientIdentity> {
+  const result = await (tauri.invoke('farming', { path })).then(value => value as ClientIdentity)
+  const farmer_identity: ClientIdentity = { public_key: result.public_key, mnemonic: result.mnemonic }
+  return farmer_identity
 }
```
Comment (nazar-pc) [inline src/lib/client.ts:215]: This should be sufficient I think:
```suggestion
export function startFarming(path: string): Promise<ClientIdentity> {
  return tauri.invoke('farming', { path }));
}
```

File: src/pages/SetupPlot.vue [hunk 6]
```diff
@@ -161,8 +161,7 @@ export default defineComponent({
     return {
       revealKey: false,
       userConfirm: false,
-      generatedPk:
-        "98da4ef21b864d2cc526dbdb2a120bd2874c36c9d0a1fb7f8c63d7f7a8b41de8f",
+      //generatedPk: "98da4ef21b864d2cc526dbdb2a120bd2874c36c9d0a1fb7f8c63d7f7a8b41de8f",
```
Comment (nazar-pc) [inline src/pages/SetupPlot.vue:164]: Time to remove?

File: src/pages/SetupPlot.vue [hunk 7]
```diff
@@ -262,9 +261,10 @@ export default defineComponent({
       })
       if (this.defaultPath != this.plotDirectory)
         await native.createDir(this.plotDirectory)
-      const public_key = await startFarming(this.plotDirectory)
+      const farmer_identity = await startFarming(this.plotDirectory)
       // TODO: find a way to store and retrieve the public key from client.ts.
-      LocalStorage.set("farmerPublicKey", public_key)
+      LocalStorage.set("farmerPublicKey", farmer_identity.public_key)
+      LocalStorage.set("mnemonic", farmer_identity.mnemonic)
```
Comment (nazar-pc) [inline src/pages/SetupPlot.vue:267]: Storing mnemonic here this way is not a great idea, but can be dealt with later as things will change anyway.

File: src-tauri/src/main.rs (hunk 1 above)
Comment (ozgunozerk) [reply src-tauri/src/main.rs:44]: Forgot to merge main into this branch... Fixing similar ones as well now

File: src/lib/client.ts (hunk 4 above)
Comment (ozgunozerk) [reply src/lib/client.ts:67]: Changing this (line 67) to `publicKey` and changing where it is used:
[src/pages](https://github.com/subspace/subspace-desktop/pull/32/files#diff-e879b9701cf110ae7a5636e48ba2803b2f6a8b808b3539e2937607f7a0200d11R265)
`farmerIdentity.public_key` into `farmerIdentity.publicKey` breaks some functionality. 
The desktop app cannot retrieve the solutions that have been found by the farmer and thus displays **0 solutions found**, although it finds solutions...

Don't know the cause yet. Also tried to change other `public_key` namings to `publicKey`, still did not work. I believe it has to match with tauri-side variable naming. Not sure though

File: src/lib/client.ts [hunk 8]
```diff
@@ -203,6 +208,6 @@ export class Client {
   }
 }
 
-export function startFarming(path: string): Promise<string> {
-  return (tauri.invoke('farming', { path })).then(value => value as string)
+export async function startFarming(path: string): Promise<ClientIdentity> {
```
Comment (nazar-pc) [inline src/lib/client.ts:211]: This will work, but function doesn't need to be `async` in this case since there is no `await` in it.

**Review by nazar-pc — APPROVED** [review]

File: src/lib/client.ts (hunk 8 above)
Comment (ozgunozerk) [reply src/lib/client.ts:211]: alright, will include in the next pr :)

**Pull request merged**